File: frontend/apps/web/src/app/core/i18n/supported-languages.ts
/**
 * The single source of truth for which languages JuggerHub ships (feature 031, FR-017/FR-018).
 * Adding a language is a one-line change here plus its catalogs / `.resx` / email folder on the
 * backend — no architecture change (SC-008). English is the source and universal fallback.
 *
 * Kept in parity with the backend supported-culture list (RequestLocalization + the language
 * `PUT` validator).
 */
export const SUPPORTED_LANGUAGES = ['en', 'de', 'es'] as const;

export type SupportedLanguage = (typeof SUPPORTED_LANGUAGES)[number];

/** The default and fallback language for the whole app (FR-018). */
export const DEFAULT_LANGUAGE: SupportedLanguage = 'en';

/** Human-readable endonym for each language — shown in the switcher in its own language (FR-014). */
export const LANGUAGE_ENDONYMS: Record<SupportedLanguage, string> = {
  en: 'English',
  de: 'Deutsch',
  es: 'Español',
};

/** transloco-locale mapping: which locale drives date/number formatting for each language (FR-009). */
export const LANG_TO_LOCALE: Record<SupportedLanguage, string> = {
  en: 'en-US',
  de: 'de-DE',
  es: 'es-ES',
};

export function isSupportedLanguage(value: string | null | undefined): value is SupportedLanguage {
  return !!value && (SUPPORTED_LANGUAGES as readonly string[]).includes(value);
}

/**
 * Collapse a BCP-47 tag to a supported base language, or null when unsupported (FR-003).
 * e.g. "de-AT" -> "de", "es-MX" -> "es", "fr" -> null.
 */
export function matchSupportedLanguage(tag: string | null | undefined): SupportedLanguage | null {
  if (!tag) return null;
  const base = tag.toLowerCase().split('-')[0];
  return isSupportedLanguage(base) ? base : null;
}

File: backend/Common/SupportedLanguages.cs
namespace JuggerHub.Common;

/// <summary>
/// The single backend source of truth for which interface languages JuggerHub supports
/// (feature 031). Kept in parity with the frontend <c>supported-languages.ts</c> constant and the
/// <c>RequestLocalization</c> supported-culture list in <c>Program.cs</c>. English is the default
/// and the universal fallback. Adding a language is a one-line change here plus its catalogs,
/// <c>.resx</c> resources and email folder — no architecture change (spec SC-008 / FR-017).
/// </summary>
public static class SupportedLanguages
{
    /// <summary>Supported BCP-47 base tags, lowercase.</summary>
    public static readonly IReadOnlyList<string> All = new[] { "en", "de", "es" };

    /// <summary>The default and universal fallback language (FR-018).</summary>
    public const string Default = "en";

    /// <summary>True when <paramref name="language"/> is one of the supported base tags.</summary>
    public static bool IsSupported(string? language) =>
        !string.IsNullOrWhiteSpace(language) && All.Contains(language.Trim().ToLowerInvariant());

    /// <summary>
    /// Collapse an arbitrary language tag (e.g. <c>"de-AT"</c>) to a supported base language,
    /// or fall back to <see cref="Default"/> when unsupported/absent (FR-003/FR-012).
    /// </summary>
    public static string ResolveOrDefault(string? tag)
    {
        if (string.IsNullOrWhiteSpace(tag))
        {
            return Default;
        }

        var baseTag = tag.Split('-')[0].Trim().ToLowerInvariant();
        return All.Contains(baseTag) ? baseTag : Default;
    }
}
